### PR TITLE
chore(http sink): Distinguish end-to-end errors

### PR DESCRIPTION
**Pull request by bruceg**
Signed-off-by: Bruce Guenter <bruce.guenter@datadoghq.com>

Ref: https://github.com/timberio/vector/pull/7325#discussion_r627817343

### CODE REVIEW

File: lib/vector-core/src/event/finalization.rs [hunk 1]
```diff
@@ -147,10 +147,14 @@ impl BatchNotifier {
 
     /// Update this notifier's status from the status of a finalized event.
     fn update_status(&self, status: EventStatus) {
-        // The status starts as Delivered and can only change to Failed
-        // here. A store cycle is much faster than fetch+update.
-        if status == EventStatus::Failed {
-            self.status.store(BatchStatus::Failed, Ordering::Relaxed);
+        // The status starts as Delivered and can only change if the new
+        // status is different than that.
+        if status != EventStatus::Delivered && status != EventStatus::Dropped {
```
Comment (lukesteensen) [inline lib/vector-core/src/event/finalization.rs:152]: nitpick: Would it make sense to put this list of variants behind a method that describes what they have in common? That could make this a little easier to scan and maybe help maintain the logic should new variants get added.

File: lib/vector-core/src/event/finalization.rs [hunk 2]
```diff
@@ -212,17 +236,18 @@ impl EventStatus {
...
-                | (Self::Delivered, Self::Dropped)
-            // Failed does not otherwise update.
-                | (Self::Failed, _) => self,
-            (Self::Delivered, _) => status,
+            // `Recorded` always overwrites existing status and is never updated
+            (_, Self::Recorded) | (Self::Recorded, _) => Self::Recorded,
+            // `Dropped` always updates to the new status.
+            (Self::Dropped, _) => status,
+            // Updates *to* `Dropped` are nonsense.
+            (_, Self::Dropped) => self,
```
Comment (lukesteensen) [inline lib/vector-core/src/event/finalization.rs:244]: If this is indicates a bug, should we panic? Or `debug_assert!(false)`.

**Review by lukesteensen — APPROVED** [review]

File: lib/vector-core/src/event/finalization.rs (hunk 1 above)
Comment (bruceg) [reply lib/vector-core/src/event/finalization.rs:152]: This optimization is a bit of a cross-domain concern. The set of statuses that can cause of an update is dependent on both the list of event statuses and the initial batch status state. So, the logic belongs in `BatchStatus`, but the only parameter is an `EventStatus`.

File: lib/vector-core/src/event/finalization.rs (hunk 2 above)
Comment (bruceg) [reply lib/vector-core/src/event/finalization.rs:244]: I would oppose a panic, as it is a harmless bug, but `debug_assert!` would make some sense to catch them in testing.

Comment (bruceg) [reply lib/vector-core/src/event/finalization.rs:244]: Hah, except then I can't test that combination in the update matrix.

**Pull request merged**